xml-conduit/src/Text/XML/Stream/Parse.hs

```diff
@@ -227,13 +227,48 @@ tokenToEvent _ es n (TokenEndElement name) =
         case n of
             []   -> (NSLevel Nothing Map.empty, [])
             x:xs -> (x, xs)
-tokenToEvent _ es n (TokenContent (ContentEntity e))
-    | Just t <- lookup e es = (es, n, [EventContent $ ContentText t])
+tokenToEvent ps es n (TokenContent (ContentEntity e))
+    = (es, n, map EventContent (resolveEntities ps es [ContentEntity e]))
 tokenToEvent _ es n (TokenContent c) = (es, n, [EventContent c])
 tokenToEvent _ es n (TokenComment c) = (es, n, [EventComment c])
 tokenToEvent _ es n (TokenDoctype t eid es') = (es ++ es', n, [EventBeginDoctype t eid, EventEndDoctype])
 tokenToEvent _ es n (TokenCDATA t) = (es, n, [EventCDATA t])
 
+resolveEntities :: ParseSettings
+                -> [(Text, Text)]  -- entity table
+                -> [Content]
+                -> [Content]
+resolveEntities ps entities = foldr go []
+ where
+  go c@(ContentEntity e) cs
+    = case expandEntity entities e of
+        Just xs -> foldr go cs xs
+        Nothing ->  c : cs
+  go c cs = c:cs
+  expandEntity es e
+    | Just t <- lookup e es =
+      case AT.parseOnly (manyTill
+                          (parseContent ps False False :: Parser Content)
+                          AT.endOfInput) t of
+        Left _      -> Nothing
+        Right xs    ->
+          -- we delete e from the entity map in resolving its contents,
+          -- to avoid infinite loops in recursive expansion:
+          case foldr (goent (filter (\(x,_) -> x /= e) es)) (Just ([], 0)) xs of
```

Comment:
I thought about doing that.  The main impediment is that it would be awkward to do this without changing the `TokenDoctype` constructor
```
| TokenDoctype Text (Maybe ExternalID) [(Text, Text)]
```
which is exported, so this would be a breaking API change. If that's okay with you, I could go ahead and make the change. Another alternative is to keep the association list in `TokenDoctype`, but convert it to a map either in `resolveEntities` or before calling `tokenToEvent` (which isn't exported and could have its signature change).

I was reluctant to make any of these changes, because I'm not really sure what the performance impact would be. I imagine most doctypes have no more than 20 internal entities, and it's quite possible that the overhead of converting the list to a map would swamp the savings from the lookup/delete.  I'm just not sure.  Is there a benchmark suite for this library?

Reply:
Fair points. There is currently no benchmark suite for this library, I might look into that after merging this pull-request.